### PR TITLE
kubeadm-token: search for existing kubeconfig files

### LINKED ISSUE #198: `kubeadm token` commands should look in standard places for kubeconfig
Currently the `kubeadm token` commands assumes it can find it's kubeconfig at `/etc/kubernetes/admin.conf`.  Instead, we should probably also look for the kubeconfig in the standard places and respect the standard env variables so that this can be used like `kubectl`.

Another option would be to move/duplicate the token management methods into `kubectl` directly as they can be managed off cluster.

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:

Add means to search the current user home path and
the environment variable KUBECONFIG for existing files if the
user does not provide a --kubeconfig flag.

If the user provides a --kubeconfig flag respect it.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubeadm/issues/198

**Special notes for your reviewer**:

i'm not 100% sure if that's the correct approach and if that's what's requested. so let's discuss it.

**Release note**:

```release-note
Search standard KubeConfig file locations when using `kubeadm token` without `--kubeconfig`.
```

### CODE REVIEW

**Comment by neolit123** [comment]
@kad @liztio

File: cmd/kubeadm/app/cmd/token.go [hunk 1]
```diff
@@ -381,3 +392,32 @@ func getClientset(file string, dryRun bool) (clientset.Interface, error) {
 	}
 	return kubeconfigutil.ClientSetFromFile(file)
 }
+
+func findExistingKubeConfig(file string) string {
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/token.go]: Most of this code exists in clientcmd.NewNonInteractiveDeferredLoadingClientConfig and should not be leveraged using the client tooling.

**Review by timothysc — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/cmd/token.go [hunk 2]
```diff
@@ -85,7 +92,7 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
 	}
 
 	tokenCmd.PersistentFlags().StringVar(&kubeConfigFile,
-		"kubeconfig", "/etc/kubernetes/admin.conf", "The KubeConfig file to use when talking to the cluster")
+		"kubeconfig", defaultKubeConfig, "The KubeConfig file to use when talking to the cluster. If the flag is not set the environment variable "+kubeConfigEnv+" and the user home directory and will be searched. If no existing KubeConfig files are found there the default location "+defaultKubeConfig+" will be used.")
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/cmd/token.go]: >If the flag is not set the environment variable "+kubeConfigEnv+" and the user home directory and will be searched.

I don't understand this. Guess the last `and` is extra?

>If no existing KubeConfig files are found there the default location "+defaultKubeConfig+" will be used."

I'm not sure if `defaultKubeConfig` can be printed in this format. Let's see if the CI will be happy.

/ok-to-test

Comment (neolit123) [reply cmd/kubeadm/app/cmd/token.go]: yes, there is an extra `and` - typo.
i will probably replace this comment based on what `clientcmd.NewNonInteractiveDeferredLoadingClientConfig` (what a name...) does exactly behind the lines. on a quick look the `cmdutil.GetJoinCommand` call might have a problem with that...

**Review by timothysc — CHANGES_REQUESTED** [review]
/assign @timothysc 
/approve 

Could you please add a UT then lgtm

**Comment by neolit123** [comment]
> Could you please add a UT then lgtm

ok, i need the other UT PR for `token` merged as a basis.

**Comment by neolit123** [comment]
link for my other PR that is needed before this:
https://github.com/kubernetes/kubernetes/pull/59220

**Comment by neolit123** [comment]
@timothysc 

update:
- added a unit test for the `NewDefaultClientConfigLoadingRules()` call using `kubeadm token delete` by setting a `KUBECONFIG` env. variable and not passing `--kubeconfig`.
- renamed some variables in `token_test.go` to make them more unique.

**Review by timothysc — APPROVED** [review]
/lgtm 

thx for the patch + update.

File: cmd/kubeadm/app/cmd/token.go [hunk 3]
```diff
@@ -50,6 +51,8 @@ import (
 	api "k8s.io/kubernetes/pkg/apis/core"
 )
 
+const defaultKubeConfig = "/etc/kubernetes/admin.conf"
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/token.go]: Could you put this in `constants.go`?

File: cmd/kubeadm/app/cmd/token.go [hunk 4]
```diff
@@ -381,3 +387,16 @@ func getClientset(file string, dryRun bool) (clientset.Interface, error) {
 	}
 	return kubeconfigutil.ClientSetFromFile(file)
 }
+
+func findExistingKubeConfig(file string) string {
+	// The user did provide a --kubeconfig flag. Respect that and threat it as an
+	// explicit path without building a DefaultClientConfigLoadingRules object.
+	if file != defaultKubeConfig {
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/token.go]: @neolit123 can you move this function to `cmd/kubeadm/app/cmd/util` please?
We have other places in kubeadm where it's used as a Kubernetes client tool, and we could/should leverage the same functionality to be consistent

**Pull request merged**